gr-fec/lib/fec_mtrx_impl.cc

```diff
@@ -396,42 +403,56 @@ gsl_matrix* fec_mtrx_impl::calc_inverse_mod2(const gsl_matrix* original_matrix)
     gsl_matrix_scale(matrix_inverse, determinant);
 
     // Take mod 2 of each element in the matrix.
-    for (row_index = 0; row_index < n; row_index++) {
-        for (col_index = 0; col_index < n; col_index++) {
+    for (unsigned int row_index = 0; row_index < n; row_index++) {
+        for (unsigned int col_index = 0; col_index < n; col_index++) {
 
-            float value = gsl_matrix_get(matrix_inverse, row_index, col_index);
+            const auto value = gsl_matrix_get(matrix_inverse, row_index, col_index);
 
             // take care of mod 2
-            int value_cast_as_int = static_cast<int>(value);
-            int temp_value = abs(fmod(value_cast_as_int, 2));
+            // FIXME: we cast to int, which will be problematic when something is
+            // x.9999998; then we do an fmod on that int; then we do an abs on the float
+            // that comes out of that, which we then truncate to int.
+            //
+            // Not many ways this could be less efficient, and more numerically
+            // questionable.
+            const int value_cast_as_int = static_cast<int>(value);
+            const int temp_value = abs(fmod(value_cast_as_int, 2));
 
             gsl_matrix_set(matrix_inverse, row_index, col_index, temp_value);
         }
     }
+    gsl_matrix_uchar* matrix_inv_gf2 = gsl_matrix_uchar_alloc(n, n);
+    unsigned char max_value = 0;
+    for (unsigned row = 0; row < n; ++row) {
+        for (unsigned col = 0; col < n; ++col) {
+            const unsigned char value = gsl_matrix_get(matrix_inverse, row, col);
+            max_value = std::max(max_value, value);
+            gsl_matrix_uchar_set(matrix_inv_gf2, row, col, value & 1);
+        }
+    }
 
-    int max_value = gsl_matrix_max(matrix_inverse);
     if (!max_value) {
         throw "Error in calc_inverse_mod2(): The matrix inverse found is all zeros.";
     }
```

Comment:
Can this even happen? Setting aside rounding problems in the above code, the only way in which I see this happening is if the inverse has come out as all zeros (or something very close to all zeros). And that should never happen. I would expect `gsl_linalg_LU_invert` to complain if it can't invert, rather than producing an all-zero matrix.

Reply:
yeah, I didn't see a case either; my guess is this is a remnant from developing the mechanism, when some mistake did lead to this, and this check actually helped identify the issue. Didn't want to remove security mechanisms, but seeing that we calculate `max_value` in vain, hm.

I think I'll leave it in for now and add a TODO to remove it when someone actually implements a "native" GF2 matrix inverse.